site/source/docs/porting/Debugging.rst

```diff
@@ -137,6 +137,58 @@ Debug printouts can even execute arbitrary JavaScript. For example::
   }
 
 
+Printing exception messages
+===========================
+
+Exceptions are thrown from WebAssembly using exception pointers, which means
+that try/catch/finally blocks in JavaScript will only receive a number, which
+represents a pointer into linear memory. In order to get the exception message,
+the user will need to create some WASM code which will extract the meaning from
+the exception. In the example code below we created a function that receives an
+``int`` which is pointer to an ``std::exception``, and by casting the pointer
+returns the ``what`` function call result.
+
+.. code-block:: cpp
+
+  #include <bind.h>
+
+  std::string getExceptionMessage(int exceptionPtr) {
+    return std::string(reinterpret_cast<std::exception *>(exceptionPtr)->what());
+  }
+
+  EMSCRIPTEN_BINDINGS(Bindings) {
+    emscripten::function("getExceptionMessage", &getExceptionMessage);
```

Comment:
So in this example what would the throw look like?  I guess it would be throwing raw char * pointer ?  e.g. `throw "mystring"` or would it be `throw new std::exception("mystring")`?

Reply:
no, it would be a thrown std::exception or subtype - for example, 
`throw std::runtime_error("foo");`
In order to handle dynamically allocated exceptions, like your example, you'll need 
`return std::string((*reinterpret_cast<std::exception **>(exceptionPtr))->what());` to handle them.

The crux of the issue is that you can only handle a single approach towards exceptions with this way.
Either you only throw statically allocated exceptions, or dynamically allocated exceptions, or strings, or integers - and each approach requires a different function to handle.

Reply:
I see.  And it sounds like it is impossible to distinguish them in JS at all because they all show up as numbers?    So you could need to call back into C++ code and use RTTI to distinguish them I guess? 

I guess for this reason is makes more sense to catch in native code where its easier to distinguish the different things one might might catch?

Reply:
Yes, to the best of my knowledge it is impossible to distinguish in JS.
I assume you could use RTTI to distinguish between the pointers, but IMO it's much better for a codebase to be consistent in the type of object it throws.

AFAIK it's not only that native code can distinguish between the different types, but also that for some types (`std::exception` and its subtypes being the prime examples) getting the error message a pointer to memory isn't sufficient, but access to the object's functions (`what()` in the case of `std::exception`) is required.